pinot-batch-ingestion/pinot-standalone/src/main/java/org/apache/pinot/ingestion/standalone/SegmentGenerationJobRunner.java

```diff
@@ -0,0 +1,318 @@
+        File localSegmentTarFile = new File(localOutputTempDir, segmentTarFileName);
+        LOGGER.info("Tarring segment from: {} to: {}", localSegmentDir, localSegmentTarFile);
+        TarGzCompressionUtils.createTarGzOfDirectory(localSegmentDir.getPath(), localSegmentTarFile.getPath());
+        long uncompressedSegmentSize = FileUtils.sizeOf(localSegmentDir);
+        long compressedSegmentSize = FileUtils.sizeOf(localSegmentTarFile);
+        LOGGER.info("Size for segment: {}, uncompressed: {}, compressed: {}", segmentName,
+            DataSize.fromBytes(uncompressedSegmentSize), DataSize.fromBytes(compressedSegmentSize));
+        //move segment to output PinotFS
+        URI outputSegmentTarURI =
+            getRelativeOutputPath(inputDirURI, inputFileURI, outputDirURI).resolve(segmentTarFileName);
+        if (!_spec.isOverwriteOutput() && outputDirFS.exists(outputSegmentTarURI)) {
+          LOGGER.warn("Not overwrite existing output segment tar file: {}", outputDirFS.exists(outputSegmentTarURI));
+        } else {
+          outputDirFS.copyFromLocalFile(localSegmentTarFile, outputSegmentTarURI);
+        }
+        FileUtils.deleteQuietly(localSegmentDir);
+        FileUtils.deleteQuietly(localSegmentTarFile);
+        FileUtils.deleteQuietly(localInputDataFile);
+      }
+    } finally {
+      //clean up
+      FileUtils.deleteDirectory(localTempDir);
+    }
+  }
+
+  private Schema getSchema() {
+    URI schemaURI;
+    try {
+      schemaURI = new URI(_spec.getTableSpec().getSchemaURI());
+    } catch (URISyntaxException e) {
+      throw new RuntimeException("Schema URI is not valid - " + _spec.getTableSpec().getSchemaURI(), e);
+    }
+    String scheme = schemaURI.getScheme();
+    String schemaJson;
+    if (PinotFSFactory.isSchemeSupported(scheme)) {
+      // Try to use PinotFS to read schema URI
+      PinotFS pinotFS = PinotFSFactory.create(scheme);
+      InputStream schemaStream;
+      try {
+        schemaStream = pinotFS.open(schemaURI);
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to open schema file stream on Pinot fs - " + schemaURI, e);
+      }
+      try {
+        schemaJson = IOUtils.toString(schemaStream, StandardCharsets.UTF_8);
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to read from schema file data stream on Pinot fs - " + schemaURI, e);
+      }
+    } else {
+      // Try to directly read from URI.
+      try {
+        schemaJson = IOUtils.toString(schemaURI, StandardCharsets.UTF_8);
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to read from Schema URI - " + schemaURI, e);
+      }
+    }
+    try {
+      return Schema.fromString(schemaJson);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to decode Pinot schema from json string - " + schemaJson, e);
+    }
+  }
+
+  private TableConfig getTableConfig() {
+    URI tableConfigURI;
+    try {
```

Comment:
Suggest get the uri string and use it once in converting to URI as well as exception string

Reply:
this is to convert the string to uri